server/mediaserver.go

```diff
@@ -281,7 +282,7 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 						monitor.StreamStarted(nonce)
 					}
 				}
-				processSegment(cxn, seg)
+				go processSegment(cxn, seg)
```

Comment:
I see that the entire function is run in a go routine now, as opposed to the old `for (true)` loop.   That seems fine.  So my comment here does not apply to this change, but it applies to the general idea of running the transcode function in a go routine at all -- 

Is it safe to spawn an arbitrary number of go routines here?  If lots of segments come in simultaneously, will it cause contention?

This looks like a job for a workpool.   I wrote a generic workpool module that might save the day, here.

https://github.com/mkrufky/genepool

Reply:
> If lots of segments come in simultaneously, will it cause contention?

Normally they should not under the "segmented live stream" assumption: a properly configured single stream should produce single segments at a comfortable cadence. Pushing RTMP streams through at wire speed (eg, via ffmpeg without `-re`) is not really a use case we should put much effort into supporting at the moment, IMO. At the very least, we'll run out of orchestrators and drop segments if that happens.

We do have the capability to absorb occasional fluctuations though; I believe the default is 8 orchestrators in the working set (one O per inflight segment).

If/when we start looking at other types of videos (eg, VOD) then we can look at building workflows specialized around that. Trying to accommodate disparate use cases under the same workflow leads to mediocrity.

> This looks like a job for a workpool.

Genepool ooks pretty nice. The B session manager is essentially a worker pool as well, albeit with some additional logic around managing Os.

Reply:
> Normally they should not under the "segmented live stream" assumption

To confirm, the issue we had before was that multiple frames were being sent at a time, not multiple segments? I believe this was the PR that fixed it: https://github.com/livepeer/lpms/pull/131

Reply:
Yes, the LPMS PR fixes an unrelated issue with multiple frames being made available from within the transcoder. @mkrufky is referring to the B receiving a deluge of segments from a stream simultaneously.